rules/go_rules.build_defs

```diff
@@ -809,6 +809,156 @@ def cgo_test(name:str, srcs:list, data:list=None, deps:list=None, visibility:lis
+        '"$TOOLS_GO" install -gcflags "-trimpath \\\"$TMP_DIR\\\"" -asmflags "-trimpath \\\"$TMP_DIR\\\"" ' + install_cmd,
+        ]
+    if CONFIG.GOROOT:
+        cmd = [f'export GOROOT={CONFIG.GOROOT}'] + cmd
+    if package_name():
+        cmd += [
+            # The outputs tend to end up in subdirectories (Go seems to match them to the location the source came from)
+            'rm -rf bin' if binary else 'rm -rf pkg',
+            'mv "$PKG_DIR/bin" .' if binary else 'mv "$PKG_DIR/pkg" .',
+        ]
+
+    if binary:
+        outs = ['bin/' + name]
+    else:
+        outs = [f'pkg/{CONFIG.GOOS}_{CONFIG.GOARCH}/{out}' for out in outs]
+        # Outputs are created one directory down from where we want them.
+        # For most it doesn't matter but the top-level one will get lost.
+        pkg_get_roots = [f'pkg/{CONFIG.GOOS}_{CONFIG.GOARCH}/{getroot}' for getroot in get_roots]
+        cmd += [f' if [ -f {out}.a ]; then mkdir -p {out} && cp {out}.a {out}; fi' for out in pkg_get_roots]
+    return build_rule(
+        name = name,
+        tag = "a_rule" if not binary else None,
+        outs = outs,
+        deps = deps,
+        srcs = srcs,
+        tools = {
+            'go': [CONFIG.GO_TOOL],
+        },
+        visibility = visibility,
+        building_description = 'Compiling...',
+        cmd = ' && '.join(cmd),
+        binary = binary,
+        requires = ['go', 'go_src'],
+        test_only = test_only,
+        labels = ['link:plz-out/go'],
+        sandbox = False,
+        needs_transitive_deps = True,
+        licences = licences,
+    )
+
+def go_mod_download(name:str, module:str, version:str, test_only:bool=False, visibility:list=None,
+                    licences:list=None, hashes:list=None, labels:list=[],_tag:str=''):
+    cmds = [
+        "export GO111MODULE=on && export TMPDIR=$TMPDIR/tmp && mkdir -p $TMPDIR",
+        "$TOOL mod init dummy", # Create a dummy module
+        f'export MOD_DIR=$($TOOL mod download -x -modcacherw -json {module}@{version} | ' 'awk -F\\" \'/"Dir": / { print $4 }\')',
+        "cp -r $MOD_DIR $OUT",
+    ]
+
+    return build_rule(
+        name = name,
+        tag = _tag,
+        outs = [f'src/{module}'],
+        tools = [CONFIG.GO_TOOL],
+        building_description = 'Fetching...',
+        cmd = ' && '.join(cmds),
+        requires = ['go'],
+        test_only = test_only,
+        labels = labels + ['link:plz-out/go'],
+        hashes = hashes,
+        sandbox = False,
+        licences = licences,
+        visibility = visibility,
+    )
+
+def go_module(name:str='', module:str='', version:str='', download:str='', deps:list=[], exported_deps:list=None,
+              visibility:list=None, test_only:bool=False, binary:bool=False, install:list=[], hashes:list=None, licences:list=None):
+    if (module or version) and download:
```

Comment:
maybe `assert` instead?

I sort of feel like one day we should get rid of `raise` and the exception types. Maybe I should add an issue.